### PR TITLE
Bugfix: Check for skipped state in task run table to fix endless duration 

### LINKED ISSUE #1110: BUG: SKIPPED states don't have end times so UI counts up duration forever
**Bug Description**

Screen Shot 2021-11-03 at 5 15 57 PM

**Steps To Reproduce**

1. Get a task run to raise a `SKIP` signal
2. Go to the Task Run page in the UI or the mapped runs page
3. Watch the duration count slowly upwards forever

**Browsers Tested:**
* [X] Chrome
* [ ] Firefox
* [ ] Safari
* [ ] Edge
* [ ] IE

----------
**Notes**

### PR DESCRIPTION
**Pull request by ThatGalNatalie**
PR Checklist:

- [ ] add a short description of what's changed to the top of the `CHANGELOG.md`
- [ ] add/update tests (or don't, for reasons explained below)

## Describe this PR
Resolves #1110

### CODE REVIEW

File: src/pages/TaskRun/MappedTaskRuns-Tile.vue [hunk 1]
```diff
@@ -274,7 +279,13 @@ export default {
           <DurationSpan
             v-if="item.start_time"
             :start-time="item.start_time"
-            :end-time="item.end_time"
+            :end-time="
+              item.end_time
+                ? item.end_time
+                : isFinished(item.state)
+                ? item.start_time
+                : null
```
Comment (pleek91) [inline src/pages/TaskRun/MappedTaskRuns-Tile.vue:287]: This ternary is copied, good opportunity to create a utility function.

File: src/pages/TaskRun/MappedTaskRuns-Tile.vue [hunk 2]
```diff
@@ -75,7 +76,11 @@ export default {
       return `%${this.searchTerm}%`
     }
   },
-  methods: {},
+  methods: {
+    isFinished(state) {
+      return FINISHED_STATES.includes(state)
+    }
```
Comment (pleek91) [inline src/pages/TaskRun/MappedTaskRuns-Tile.vue:82]: let's make this a utility function in an external file so it can be reused in both components.

**Review by pleek91 — DISMISSED** [review]
Everything looks good. But lets DRY up the logic by creating some utility functions.

**Review by zhen0 — DISMISSED** [review]
This looks good to me but I also see that duration count in the task run details tile.  Could we add your durationCalc method there too?
Screen Shot 2021-11-30 at 1 14 12 PM

File: src/pages/Task/TaskRunTable-Tile.vue [hunk 3]
```diff
@@ -71,7 +72,11 @@ export default {
       return `%${this.searchTerm}%`
     }
   },
-  methods: {},
+  methods: {
+    durationCalc(startTime, endTime, state) {
```
Comment (stackoverfloweth) [inline src/pages/Task/TaskRunTable-Tile.vue:76]: nit: I would import the method with this full name `calculateDuration`, and rather than using another name here but calling the imported method I would just make it available to the template with

```
methods: {
    calculateDuration
}
```

which should work fine since the arguments do not change

File: src/pages/TaskRun/MappedTaskRuns-Tile.vue [hunk 4]
```diff
@@ -75,7 +76,11 @@ export default {
       return `%${this.searchTerm}%`
     }
   },
-  methods: {},
+  methods: {
+    durationCalc(startTime, endTime, state) {
```
Comment (stackoverfloweth) [inline src/pages/TaskRun/MappedTaskRuns-Tile.vue:80]: same suggestion here

File: src/utils/calculateDuration.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { FINISHED_STATES } from '@/utils/states'
+
+function isFinished(state) {
+  return FINISHED_STATES.includes(state)
+}
+
+export function duration(startTime, endTime, state) {
```
Comment (stackoverfloweth) [inline src/utils/calculateDuration.js:7]: I would recommend either naming the function `calculateDuration`, and/or consider having it be the default export (which would require removing curly brackets where you're importing it)

File: src/utils/calculateDuration.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+  return endTime ? endTime : isFinished(state) ? startTime : null
```
Comment (stackoverfloweth) [inline src/utils/calculateDuration.js:8]: I would be in favor or more lines over nested ternaries. IMO This can be pretty hard to follow compared to 

```
if (endTime) {
   return endTime
}

if (isFinished(state)) {
    return startTime
}

return null
```

File: src/utils/calculateDuration.js (hunk 5 above)
Comment (pleek91) [reply src/utils/calculateDuration.js:7]: Generally a utilities file wouldn't have a default export. So I'd go with the more clear name route. 

However I'd suggest this utility belongs in the `@utils/states.js` since it has to do with states. a utility file name `calculateDuration` is probably too specific.

File: src/utils/calculateDuration.js (hunk 6 above)
Comment (pleek91) [reply src/utils/calculateDuration.js:8]: +1

File: src/utils/calculateDuration.js (hunk 5 above)
Comment (ThatGalNatalie) [reply src/utils/calculateDuration.js:7]: I think putting it in @utils/states.js makes sense too

**Review by zhen0 — APPROVED** [review]
Thanks @ThatGalNatalie
